pkg/build/controller/build/build_controller.go

```diff
@@ -1588,23 +1607,153 @@ func (bc *BuildController) handleControllerConfig() error {
+	}
+	return additionalCA.Data, nil
+}
+
+func (bc *BuildController) createBuildRegistriesConfigData(config *configv1.Build) (string, error) {
+	registriesConfig := config.Spec.BuildDefaults.RegistriesConfig
+	if registriesConfig.SearchRegistries == nil && len(registriesConfig.InsecureRegistries) == 0 {
+		glog.V(4).Info("using default search and insecure registry settings for builds")
+		return "", nil
+	}
+	configObj := tomlConfig{
+		Registries: registries{
+			// Default search to docker.io
+			Search: registryList{
+				Registries: []string{"docker.io"},
+			},
+			Insecure: registryList{
+				Registries: registriesConfig.InsecureRegistries,
+			},
+		},
+	}
+	// Override search if provided SearchRegistries is non-nil
+	// This allows empty lists, which disables image search for builds
+	if registriesConfig.SearchRegistries != nil {
+		configObj.Registries.Search.Registries = *registriesConfig.SearchRegistries
+	}
+
+	configTOML, err := toml.Marshal(configObj)
+	if err != nil {
+		return "", err
+	}
+	if len(configTOML) == 0 {
+		glog.V(4).Info("using default search and insecure registry settings for builds")
+		return "", nil
+	}
+	glog.V(4).Info("overrode search and insecure registry settings for builds")
+	glog.V(5).Infof("generated registries.conf for build pods: \n%s", string(configTOML))
+	return string(configTOML), nil
+}
+
+func (bc *BuildController) createBuildSignaturePolicyData(config *configv1.Build) (string, error) {
+	registriesConfig := config.Spec.BuildDefaults.RegistriesConfig
+	if len(registriesConfig.AllowedRegistries) == 0 && len(registriesConfig.BlockedRegistries) == 0 {
+		glog.V(4).Info("allowing builds to pull images from all registries")
+		return "", nil
+	}
+	if len(registriesConfig.AllowedRegistries) != 0 && len(registriesConfig.BlockedRegistries) != 0 {
+		return "", fmt.Errorf("invalid registries config: only one of AllowedRegistries or BlockedRegistries may be specified")
+	}
+	policyObj := &signature.Policy{}
+	dockerScopes := make(signature.PolicyTransportScopes)
+	if len(registriesConfig.AllowedRegistries) > 0 {
+		glog.V(4).Infof("only allowing image pulls from %s for builds", registriesConfig.AllowedRegistries)
+		policyObj.Default = signature.PolicyRequirements{
+			signature.NewPRReject(),
+		}
+		for _, registry := range registriesConfig.AllowedRegistries {
+			dockerScopes[registry] = signature.PolicyRequirements{
+				signature.NewPRInsecureAcceptAnything(),
+			}
+		}
+	}
+	if len(registriesConfig.BlockedRegistries) > 0 {
+		glog.V(4).Infof("blocking image pulls from %s for builds", registriesConfig.BlockedRegistries)
+		policyObj.Default = signature.PolicyRequirements{
+			signature.NewPRInsecureAcceptAnything(),
+		}
+		for _, registry := range registriesConfig.BlockedRegistries {
+			dockerScopes[registry] = signature.PolicyRequirements{
+				signature.NewPRReject(),
+			}
+		}
+	}
+	policyObj.Transports = map[string]signature.PolicyTransportScopes{
+		"docker": dockerScopes,
```

Comment:
@nalind @rhatdan do we need other scopes (atomic, oci)?

Reply:
Not `oci:` (that’s for on-disk OCI archives/directories).

As for `atomic`, in principle, in the `BlockedRegistries` case, yes: The `atomic:` transport is deprecated, and neither `docker pull` nor `atomic pull` use it AFAICS. It would only matter if it were possible to specify an image using the `atomic:some.hostname.com/image` format; AFAICS CRI-O might be able accept that syntax, I don’t know whether it’s possible for a user to cause such an image name to be used by Kubelet / the rest of OpenShift.

Reply:
At this level, those names correspond to different transports, with "docker" corresponding to pulling from and pushing to image registries like Docker Hub.  There's a brief rundown of some of the others in the skopeo(1) man page.

Of those, the ones that don't end up with us directly manipulating files on disk are "docker" (a registry), "atomic" (an imagestream, in the form "`atomic:`hostname/namespace/stream:tag"), and "docker-daemon" (a docker daemon, accessed via its REST API).

Unless we expect to be importing images from a docker daemon running on the node (the location and method of contacting the daemon are either taken from hard-coded defaults, or fields in a `SystemContext` that we control and don't set), I don't think we need to care about "docker-daemon".  I don't have much experience with the "atomic" transport, but since we have a reliable hostname by which processes in the cluster can find the registry, we might want to shut that down preemptively.

In `pullDaemonlessImage()` we're forcing "docker://" as the transport when we pass an image name to `buildah.Pull()` and in the image destination when we call `buildah.Push()`, but there is still an implicit path when we hit a `FROM` instruction, so blocking "atomic" might be prudent.